lightpath/controller.py

```diff
@@ -119,21 +122,42 @@ def load_beamline(self, endstation: str):
         for branch in end_branches:
             # Find the paths from each source to the desired line
             for src in self.sources:
-                if nx.has_path(self.graph, src, branch):
-                    found_paths = nx.all_simple_paths(self.graph,
-                                                      source=src,
-                                                      target=branch)
-                    paths.extend(found_paths)
-                else:
-                    logger.debug(f'No path between {src} and {branch}')
+                try:
+                    if nx.has_path(self.graph, src, branch):
+                        found_paths = nx.all_simple_paths(self.graph,
+                                                          source=src,
+                                                          target=branch)
+                        paths.extend(found_paths)
+                    else:
+                        logger.debug(f'No path between {src} and {branch}')
+                except NodeNotFound:
+                    logger.debug(f'Either source {src} or target {branch} '
+                                 'not found.')
 
         self.beamlines[endstation] = []
         for path in paths:
             subgraph = self.graph.subgraph(path)
             devices = [dat['dev'] for _, dat in subgraph.nodes.data()
                        if dat['dev'] is not None]
             bp = BeamPath(*devices, name=endstation)
-            self.beamlines[endstation].append(bp)
+
+            if isinstance(end_branches, dict):
```

Comment:
This seems worthy of pulling out into a utility function (`end_branch_to_last_z(end_branches, path[-1])` or something)

Reply:
I think I tend to only pull things out into a utility function if I use it more than once, but it would make things more readable.

Reply:
Yeah, I like the readability of utility functions - I've tended to put more in nowadays than I used to, and I think my code style has improved because of it.